### ISSUE TITLE
Recall issues with dataset compared to CatBoost

### ISSUE THREAD

**Issue by hexgnu** [microsoft/interpret#69]
Hey there, I'm seeing some major differences between CatBoost and EBMs in my work.

I was able to see a result of:

```
              precision    recall  f1-score   support

         0.0       0.89      0.86      0.88      3650
         1.0       0.94      0.96      0.95      8816

    accuracy                           0.93     12466
   macro avg       0.92      0.91      0.91     12466
weighted avg       0.93      0.93      0.93     12466
```

On CatBoost but a result of below on the EBMs
```
              precision    recall  f1-score   support

           0       1.00      0.00      0.00      3542
           1       0.50      1.00      0.67      3542

    accuracy                           0.50      7084
   macro avg       0.75      0.50      0.33      7084
weighted avg       0.75      0.50      0.33      7084
```

I have tried balancing the dataset and doing some sampling strategies but nothing seems to yield the same results on my dataset. Any help is appreciated thank you!

-Matt

**Comment by interpret-ml** [comment]
Hi @hexgnu, 

Thanks for bringing this up! We're having a bit of trouble reproducing this -- we just tried to run it on the Adult income dataset, and the results between EBM and CatBoost are very similar:

![image](https://user-images.githubusercontent.com/50031429/65926773-bfe6d680-e3ab-11e9-97fd-b0e447b93fe5.png)

In your classification metrics, it seems like the classes have different supports (ex: CatBoost's report has 12,466 total instances and EBM's only has 7084). Is it possible that EBM is being evaluated on a different dataset? 


Here's the code to reproduce our results on the Adult Income dataset:

```python

import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
    header=None)
df.columns = [
    "Age", "WorkClass", "fnlwgt", "Education", "EducationNum",
    "MaritalStatus", "Occupation", "Relationship", "Race", "Gender",
    "CapitalGain", "CapitalLoss", "HoursPerWeek", "NativeCountry", "Income"
]
# df = df.sample(frac=0.1, random_state=1)
train_cols = df.columns[0:-1]
label = df.columns[-1]
X = df[train_cols]
y = df[label].apply(lambda x: 0 if x == " <=50K" else 1) #Turning response into 0 and 1

seed = 1
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=seed)


from interpret.glassbox import ExplainableBoostingClassifier

ebm = ExplainableBoostingClassifier(random_state=seed)
ebm.fit(X_train, y_train)

from catboost import CatBoostClassifier

cat = CatBoostClassifier(cat_features=np.where(X_train.dtypes == np.object)[0], silent=True) 
cat.fit(X_train, y_train)

from sklearn.metrics import classification_report


print("EBM Results: \n\n")
print(classification_report(y_test, ebm.predict(X_test)))
      
print("CatBoost Results: \n\n")
print(classification_report(y_test, cat.predict(X_test)))

**Comment by hexgnu** [comment]
I was trying out a few things here's the results with the same exact dataset:

```
print(classification_report(y_test, ebm.predict(X_test)))

              precision    recall  f1-score   support

           0       1.00      0.00      0.00      3542
           1       0.72      1.00      0.83      8924

    accuracy                           0.72     12466
   macro avg       0.86      0.50      0.42     12466
weighted avg       0.80      0.72      0.60     12466
```

This is trained on the same training data. Maybe something to do with weighting of the training instances?

When I do the adult dataset (catboost vs this ebm) I get similar results. 

I think the major difference I am seeing is that the dataset I am using is incredibly sparse. So maybe it has something to do with weighting of instances in training?

**Comment by hexgnu** [comment]
Ah! it is a sparsity issue.

So when I remove columns with super high sparsity (i.e. only 1 case of a number != nan). The recall and precision issues go away quickly.

Is there any sort of work being done about removing columns with very little feature importance out of the model?

**Comment by hexgnu** [comment]
So digging in I see that there are quite a few columns I have which have less than 1 bit of entropy. Which means that if we fed this into this GAM model it'd be calculating f(no_entropy) = mean_log_odds I'm assuming. I would assume this really throws things off.

What is the appetite for pruning columns in the EBM? I can prune them before hand and fix my own problem and get this working but... wondering if there's a way to trim columns without any entropy in this work.

Since CatBoost is a tree I think it just ends up ignoring those inputs which are irrelevant.

Thanks!

**Comment by interpret-ml** [comment]
Hi @hexgnu, 

Thanks for your detailed investigative work! We're still having some troubles replicating similar precision/recall differences on our side, even with pathologically synthetic data that's full of similar problems (i.e. many sparse features). 

Is it possible you could share your data, or further characterize it in some way? Alternatively, if you can help us make a synthetic dataset that has the same problems, we'd be able to debug this much more easily. Really appreciate all the help!

-InterpretML Team

**Comment by hexgnu** [comment]
Hrmmm weird now I can't reproduce. I'll reopen if I can. Thanks for the response :)
